src/transactions/contracts/HostVal.cpp

```diff
@@ -0,0 +1,676 @@
+    {
+        return eqOther(a);
+    }
+
+    bool
+    operator()(Price const& p) const
+    {
+        return eqOther(p);
+    }
+
+    bool
+    operator()(AccountID const& a) const
+    {
+        return eqOther(a);
+    }
+};
+
+bool
+HostValInContext::operator==(HostValInContext const& other) const
+{
+    ZoneScoped;
+    releaseAssert(mCtx);
+    releaseAssert(mCtx == other.mCtx);
+    if (mVal.payload() == other.mVal.payload())
+    {
+        // Fast path: bitwise identical.
+        return true;
+    }
+    if (mVal.isObject() && other.mVal.isObject())
+    {
+        releaseAssert(mCtx);
+        std::unique_ptr<HostObject const> const& ptr = mCtx->getObject(mVal);
+        std::unique_ptr<HostObject const> const& otherPtr =
+            mCtx->getObject(other.mVal);
+        if (!ptr || !otherPtr)
+        {
+            // Sub-fast-path: one or the other nullptr.
+            return !ptr && !otherPtr;
+        }
+        // Catch unequal-variants first.
+        auto idx = ptr->index();
+        auto otherIdx = otherPtr->index();
+        if (idx != otherIdx)
+        {
+            return false;
+        }
+        // Deep compare same-variant objects.
+        EqHostObjectVisitor ehov{*otherPtr};
+        return std::visit(ehov, *ptr);
+    }
+    // Anything that wasn't shallow-identical and isn't an object
+    // isn't equal.
+    return false;
+}
+
+struct LtHostObjectVisitor
+{
+    HostObject const& mOther;
+
+    template <typename T>
+    bool
+    ltOther(T const& v) const
+    {
+        releaseAssert(std::holds_alternative<T>(mOther));
+        return v < std::get<T>(mOther);
+    }
+
+    template <typename T>
+    bool
+    ltOtherPtr(xdr::pointer<T> const& p) const
+    {
+        releaseAssert(std::holds_alternative<xdr::pointer<T>>(mOther));
+        auto const& otherP = std::get<xdr::pointer<T>>(mOther);
+        if (p.get() == otherP.get())
+        {
+            return false;
+        }
+        if (p && !otherP)
+        {
+            return false;
+        }
+        else if (!p && otherP)
+        {
+            return true;
+        }
+        else if (!p && !otherP)
+        {
+            return false;
+        }
+        else
+        {
+            return *p < *otherP;
+        }
+    }
+
+    bool
+    operator()(HostBox const& box) const
+    {
+        return ltOther(box);
+    }
+
+    bool
+    operator()(HostVec const& vec) const
+    {
+        releaseAssert(std::holds_alternative<HostVec>(mOther));
+        auto const& otherVec = std::get<HostVec>(mOther);
+        return std::lexicographical_compare(vec.begin(), vec.end(),
+                                            otherVec.begin(), otherVec.end());
+    }
+
+    bool
+    operator()(HostMap const& map) const
+    {
+        releaseAssert(std::holds_alternative<HostMap>(mOther));
+        auto const& otherMap = std::get<HostMap>(mOther);
+        return std::lexicographical_compare(map.begin(), map.end(),
```

Comment:
Per https://sinusoid.es/immer/containers.html#map, "Immutable unordered mapping of values from type K to type T." Given the "unordered", does this always work?